### DISCUSSION TITLE
Should TS features be included with the bundles for use in Deno?

### DISCUSSION THREAD

**Discussion by bicknellr** [lit/lit#2728] [Ideas]
This discussion is meant to expand on [this thread in the Lit Slack](https://lit-and-friends.slack.com/archives/C03PF4L4L/p1649860843324719).

Deno's built-in support for TypeScript makes it possible to write and run TypeScript without a distinct build step. Should Deno + TS without other tools be considered a target workflow for the bundles? If so, should the bundles include features meant strictly for TypeScript? (i.e. type declarations, Lit's decorators)

cc @duncanmak @taylor-vann

**Comment by taylor-vann** [comment]
I think there's a lot of love for webcomponents in Deno and it's a great environment for new ideas that are browser focused, Lit could really make an impact. Deno encourages shallow dependencies, something where Lit shines. I've already used Lit with success in deno, direct support would be even better!

**Comment by zandaqo** [comment]
I use Deno for library development, including for [compago](https://github.com/zandaqo/compago) (framework/state management lib for Lit). Importing Lit did prove to be tricky due to issues with typings in modules generated by Skypack and ESM, so having Lit itself support Deno would be a boon.

Deno provides unparalleled DX for development and I recommend to check it out if you're not going to deploy on it/use it as a runtime. A heads-up, though, when it comes to Lit, there is an [issue](https://github.com/dprint/dprint-plugin-typescript/issues/9) with the Deno formatter not (officially) supporting html template tags.

**Reply by taylor-vann** [reply]
I completely agree about the unparalleled DX. It's been incredibly difficult to work in NPM after some time in deno <3

**Reply by beingminimal** [reply]
Hey we also want to use deno and lit in production.
But we have hardly found any sample repo for that and while we were searching we have found this also.
See this long awaited bug fix https://github.com/denoland/deno/issues/13048
Due to such things now it is confusing for developers, how to solve such an issue to run lit in deno?
And we will be so grateful to you if you can solve this for us.
Awaiting your response. Thanks

**Reply by zandaqo** [reply]
For bundling my lit apps, I use [esbuild](https://esbuild.github.io) (with a [plugin](https://github.com/zandaqo/esbuild-plugin-lit)). 

I haven't had much experience with the Deno bundler--when I first encountered it, there was even talk about discontinuing it in favor of esbuild, but they decided against it, so I assume in time they are going to fix this and other issues.

**Comment by taylor-vann** [comment]
@zandaqo @codecycleteam 
Lit almost works out of the box with skypack :D
And i think it will work without all the extra types imports below once https://github.com/lit/lit/issues/2383 is resolved

When I create projects in deno with Lit:

1) set tsconfig to include the DOM
```
{
  "compilerOptions": {
    "lib": ["dom", "esnext"],
    "strict": true,
    "experimentalDecorators": true
  }
}
```

2) make sure `TrustedTypes` is available.

```
// deps.ts

import "https://cdn.skypack.dev/trusted-types?dts";

export type { TemplateResult } from "https://cdn.skypack.dev/lit?dts";

// @deno-types="https://cdn.skypack.dev/lit?dts"
export { css, html, LitElement } from "https://cdn.skypack.dev/lit";
export {
  customElement,
  property,
  state,
} from "https://cdn.skypack.dev/lit/decorators";
```

3) run the deno bundle command

```
deno bundle --config ./tsconfig.json ./mod.ts ./bundle.js
```

**Reply by zandaqo** [reply]
@taylor-vann Have you tried this code with dnt? It's having hard time figuring out Skypacks `?dts`, that's why I have been using ESM:
deps.ts:
```ts
import "https://cdn.skypack.dev/trusted-types?dts";

export {
  Directive,
  directive,
  PartType,
} from "https://cdn.skypack.dev/lit@2.2.1/directive.js";

export type {
  EventPart,
  PartInfo,
} from "https://cdn.skypack.dev/lit@2.2.1/directive.js?dts";
```
mappings:
```js
  mappings: {
    "https://cdn.skypack.dev/lit@2.2.1": {
      name: "lit",
      version: "^2.2.1",
    },
    "https://cdn.skypack.dev/lit@2.2.1/directive.js": {
      name: "lit",
      version: "^2.2.1",
      subPath: "directive.js",
    },
    "https://cdn.skypack.dev/trusted-types": {
      name: "trusted-types",
      version: "^2.0.0",
    },
  }
```
```
error: Uncaught (in promise) Specifier https://cdn.skypack.dev/lit@2.2.1 with version ^2.2.1 did not match specifier https://cdn.skypack.dev/lit@2.2.1/directive.js?dts with version 2.2.1.
```

**Reply by taylor-vann** [reply]
sorry for the late response!
i'll take a look at this i don't use import maps or mappings but I'll give it a shot!

**Reply by tdillon** [reply]
@taylor-vann, when you setup a new Deno project as you outlined above, do reactive properties work?

**Reply by taylor-vann** [reply]
Yes :) I can confirm reactive properties work.
Here is an example of lit (and lit's decorators) being used to represent router state
https://github.com/herebythere/router/tree/main/docs

**Reply by tdillon** [reply]
@taylor-vann I agree the reactive properties work during the initial render, but the experience I'm having is subsequent updates to the reactive properties are not working.

Here is my super simple Deno example that doesn't work: https://github.com/tdillon/test-lit-reactive-properties.

I also have a [Lit Playground example](https://lit.dev/playground/#project=W3sibmFtZSI6Im1haW4udHMiLCJjb250ZW50IjoiaW1wb3J0IHtodG1sLCBMaXRFbGVtZW50fSBmcm9tICdsaXQnO1xuaW1wb3J0IHtjdXN0b21FbGVtZW50LCBwcm9wZXJ0eX0gZnJvbSAnbGl0L2RlY29yYXRvcnMuanMnO1xuXG5AY3VzdG9tRWxlbWVudCgnbXktYXBwJylcbmV4cG9ydCBjbGFzcyBBcHAgZXh0ZW5kcyBMaXRFbGVtZW50IHtcbiAgICBAcHJvcGVydHkoKVxuICAgIGZvbyA9ICc_Pz8nO1xuXG4gICAgcmVuZGVyKCkge1xuICAgICAgICByZXR1cm4gaHRtbGA8aDE-TXkgQXBwOiAke3RoaXMuZm9vfTwvaDE-YFxuICAgIH1cbn1cblxuY29uc3QgbXlBcHAgPSBuZXcgQXBwKClcblxuZG9jdW1lbnQuYm9keS5hcHBlbmRDaGlsZChteUFwcClcblxuZG9jdW1lbnQuYm9keS5hZGRFdmVudExpc3RlbmVyKCdjbGljaycsIGUgPT4ge1xuICAgIGNvbnNvbGUubG9nKCdiZWZvcmUnLCBteUFwcC5mb28sIG15QXBwLmdldEF0dHJpYnV0ZSgnZm9vJykpXG4gICAgbXlBcHAuZm9vID0gJ2JvYidcbiAgICBteUFwcC5zZXRBdHRyaWJ1dGUoJ2ZvbycsICdib2InKVxuICAgIGNvbnNvbGUubG9nKCdhZnRlcicsIG15QXBwLmZvbywgbXlBcHAuZ2V0QXR0cmlidXRlKCdmb28nKSlcbn0pXG4ifSx7Im5hbWUiOiJpbmRleC5odG1sIiwiY29udGVudCI6IjwhRE9DVFlQRSBodG1sPlxuPHNjcmlwdCB0eXBlPVwibW9kdWxlXCIgc3JjPVwiLi9tYWluLmpzXCI-PC9zY3JpcHQ-XG4ifSx7Im5hbWUiOiJwYWNrYWdlLmpzb24iLCJjb250ZW50Ijoie1xuICBcImRlcGVuZGVuY2llc1wiOiB7XG4gICAgXCJsaXRcIjogXCJeMi4wLjBcIixcbiAgICBcIkBsaXQvcmVhY3RpdmUtZWxlbWVudFwiOiBcIl4xLjAuMFwiLFxuICAgIFwibGl0LWVsZW1lbnRcIjogXCJeMy4wLjBcIixcbiAgICBcImxpdC1odG1sXCI6IFwiXjIuMC4wXCJcbiAgfVxufSIsImhpZGRlbiI6dHJ1ZX1d) that works.  Just click the `My App: ???` text and it will update the reactive property.

**Reply by taylor-vann** [reply]
Hey sorry for the late reply!
I'll take a look at your demo :D thanks
This issue isn't prioritized right now. But I'll keep working on this and share updates as they come!

**Reply by taylor-vann** [reply]
[edited, saw the deno.json]

Looking at my example though, i might be setting a property once aaaand might not ever be updating it! x_x
I'll double check.

**Comment by zandaqo** [comment]
@tdillon @taylor-vann I looked into it a bit more:
1. Using trusted types and adding `?dts` to the imports indeed fixes importing as you can see in my [library](https://github.com/zandaqo/compago/blob/b9bbe7ff0f986b04d440c33612e18497ab590f90/deps.ts).
2. The reason reactive properties are not working has to do with the infamous [overriding of class fields](https://lit.dev/docs/components/properties/#avoiding-issues-with-class-fields). Deno by default [sets](https://github.com/denoland/deno/issues/12393#issuecomment-940576147) `useDefineForClassFields` to `true`. I'm not yet sure how to go around this without changing the component code and separating declaration and assignment of default properties.
